src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java

```diff
@@ -740,43 +654,41 @@ protected Future<Map<String, List<Piece>>> processItemsUpdate(
     });
   }
 
-  protected Future<Void> processHoldingsUpdate(Map<String, Map<String, Location>> pieceLocationsGroupedByPoLine,
-      Map<String, List<Piece>> piecesGroupedByPoLine, PoLineAndTitleById poLinesAndTitlesById, RequestContext requestContext) {
-    List<Future<Boolean>> futuresForHoldingsUpdates = new ArrayList<>();
-    StreamEx.ofValues(piecesGroupedByPoLine)
-      .flatMap(List::stream)
-      .forEach(piece -> {
-        CompositePoLine poLine = poLinesAndTitlesById.poLineById.get(piece.getPoLineId());
-        if (poLine == null) {
-          logger.error("POLine associated with piece '{}' cannot be found", piece.getId());
-          addError(piece.getPoLineId(), piece.getId(), ITEM_UPDATE_FAILED.toError());
-          return;
-        }
-        Title title = poLinesAndTitlesById.titleById.get(piece.getTitleId());
-        if (title == null) {
-          logger.error("Piece with id {} : title with id {} was not found within the po line titles.",
-            piece.getId(), piece.getTitleId());
-          addError(piece.getPoLineId(), piece.getId(), ITEM_UPDATE_FAILED.toError());
-          return;
-        }
-        Location receivedPieceLocation = pieceLocationsGroupedByPoLine.get(poLine.getId()).get(piece.getId());
+  /**
+   * @param item  inventory item
+   * @param piece piece associated with the item
+   * @return future indicating if the item update is successful.
+   */
+  protected abstract Future<Boolean> receiveInventoryItemAndUpdatePiece(JsonObject item, Piece piece, RequestContext locationContext);
 
-        if (holdingUpdateOnCheckinReceiveRequired(piece, receivedPieceLocation, poLine)) {
-          futuresForHoldingsUpdates.add(createHoldingsForChangedLocations(piece, title.getInstanceId(),
-            receivedPieceLocation, requestContext));
-        }
-      });
+  private boolean holdingUpdateOnCheckinReceiveRequired(Piece piece, CompositePoLine poLine) {
+    if (piece.getFormat() == Piece.Format.ELECTRONIC && !PoLineCommonUtil.isHoldingUpdateRequiredForEresource(poLine)) {
+      return false;
+    }
+    if (piece.getFormat() != Piece.Format.ELECTRONIC && !PoLineCommonUtil.isHoldingUpdateRequiredForPhysical(poLine)) {
```

Comment:
I think we don't need apply this approach for all places, because inverse conditions also hard to read.
Previous implementation looked more straightforward than this line:
if (piece.getFormat() != Piece.Format.ELECTRONIC && !PoLineCommonUtil.isHoldingUpdateRequiredForPhysical(poLine))

Reply:
BTW there is a method `PoLineCommonUtil::isHoldingsUpdateRequired` and it uses a bit different logic:
`return isHoldingUpdateRequiredForPhysical(compPOL) || isHoldingUpdateRequiredForEresource(compPOL)`
Do you know why we have in fact same check but different implementation? Maybe this is a bug...